beacon-chain/sync/data_columns.go

```diff
@@ -174,7 +237,316 @@ func RequestDataColumnSidecarsByRoot(
 	}
 
 	// If we still have remaining columns after all retries, return error
-	return nil, errors.Errorf("failed to retrieve all requested data columns after retries for block root=%#x, missing columns=%v", blockRoot, uint64MapToSortedSlice(remainingMissingColumns))
+	return nil, fmt.Errorf("failed to retrieve all requested data columns after retries for block root=%#x, %w", blockRoot, NewUnavailableColumnsError(uint64MapToSortedSlice(remainingMissingColumns)))
+}
+
+func verifyColumnsForBlock(block blocks.ROBlock, columns []blocks.RODataColumn, newColumnsVerifier verification.NewDataColumnsVerifier) ([]blocks.VerifiedRODataColumn, error) {
+	// Check if returned data columns align with the block.
+	if err := verify.DataColumnsAlignWithBlock(block, columns); err != nil {
+		return nil, errors.Wrap(err, "align with block failed")
+	}
+
+	// Verify the received sidecars.
+	verifier := newColumnsVerifier(columns, verification.ByRootRequestDataColumnSidecarRequirements)
+
+	if err := verifier.Valid(); err != nil {
+		return nil, errors.Wrap(err, "valid verification failed")
+	}
+
+	if err := verifier.SidecarInclusionProven(); err != nil {
+		return nil, errors.Wrap(err, "sidecar inclusion proof verification failed")
+	}
+
+	if err := verifier.SidecarKzgProofVerified(); err != nil {
+		return nil, errors.Wrap(err, "sidecar KZG proof verification failed")
+	}
+
+	// Upgrade the sidecars to verified sidecars.
+	verifiedColumns, err := verifier.VerifiedRODataColumns()
+	if err != nil {
+		// This should never happen.
+		return nil, errors.Wrap(err, "verified data columns")
+	}
+
+	return verifiedColumns, nil
+}
+
+// ReconstructDataColumnsByRoot attempts to reconstruct a requested set of data columns for a given block.
+// It identifies available columns from peers, calculates the recovery threshold, fetches a set of 'recoveryThreshold' columns.
+// Once a fetch succeeds, it uses erasure coding to reconstruct the full set of columns and returns
+// the originally requested ones. Errors during fetch or reconstruction are fatal.
+func ReconstructDataColumnsByRoot(
+	ctx context.Context,
+	dataColumnsToFetch []uint64,
+	block blocks.ROBlock,
+	peers []core.PeerID,
+	clock *startup.Clock,
+	p2p p2p.P2P,
+	ctxMap ContextByteVersions,
+	newColumnsVerifier verification.NewDataColumnsVerifier,
+) ([]blocks.VerifiedRODataColumn, error) {
+	if len(dataColumnsToFetch) == 0 {
+		return nil, nil // Nothing requested, nothing to recover.
+	}
+
+	// Determine recovery threshold.
+	numberOfColumns := int(params.BeaconConfig().NumberOfColumns)
+	// If total is odd, then we need total / 2 + 1 columns to reconstruct.
+	// If total is even, then we need total / 2 columns to reconstruct.
+	recoveryThreshold := numberOfColumns/2 + numberOfColumns%2
```

Comment:
It's not totally true (see my other comment): If you already fetched (via `OnlyRequest...`) some columns, you need to fetch less than 64 extra columns to reach your goal.

Reply:
[62572a8](https://github.com/OffchainLabs/prysm/pull/15107/commits/62572a8fbb815d9a4eae6f5ad16758db99f05c40)